### ISSUE TITLE
runtime: SIGPROF mishandled on C thread

### ISSUE THREAD

**Issue by ianlancetaylor** [golang/go#15994]
In a cgo-using program a `SIGPROF` signal that arrives on a thread started by C code will call the following sequence of functions in the runtime package:

    cgoSigtramp
    sigtramp
    sigtrampgo
    sigfwdgo (which will normally return false)
    badsignal
    cgocallback
    badsignalgo
    raisebadsignal

`raisebadsignal` will observe that it got a `SIGPROF` signal and return, and the whole call stack will unwind having done nothing.

However, the call to `cgocallback` hides a set of internal calls, that include waiting for a new thread to be available to run the Go function `badsignalgo`.  Waiting for a new thread can in some cases mean starting a new thread.  Starting a new thread in a cgo-using program means calling `_cgo_thread_start`, which calls `malloc`.

In other words, we've called `malloc` in a signal handler.  That is not a good idea in general.  In practice, if the `SIGPROF` occurred while executing within `malloc`, as is not unlikely, the best result we can hope for is a program deadlock.

We need to, at the very least, check for `SIGPROF` before calling `cgocallback`.  No other non-forwarded signal is particularly likely to occur during `malloc`, so that may be sufficient for 1.7.
